ddsrouter_core/src/cpp/writer/implementations/rtps/MultiWriter.cpp

```diff
@@ -0,0 +1,157 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file MultiWriter.cpp
+ */
+
+#include <fastrtps/rtps/RTPSDomain.h>
+#include <fastrtps/rtps/participant/RTPSParticipant.h>
+#include <fastrtps/rtps/common/CacheChange.h>
+
+#include <ddsrouter_utils/exception/InitializationException.hpp>
+#include <ddsrouter_utils/Log.hpp>
+#include <efficiency/cache_change/CacheChangePool.hpp>
+#include <writer/implementations/rtps/MultiWriter.hpp>
+#include <writer/implementations/rtps/QoSSpecificWriter.hpp>
+#include <writer/implementations/rtps/filter/RepeaterDataFilter.hpp>
+#include <writer/implementations/rtps/filter/SelfDataFilter.hpp>
+#include <types/dds/RouterCacheChange.hpp>
+
+namespace eprosima {
+namespace ddsrouter {
+namespace core {
+namespace rtps {
+
+using namespace eprosima::ddsrouter::core::types;
+
+MultiWriter::MultiWriter(
+        const ParticipantId& participant_id,
+        const DdsTopic& topic,
+        std::shared_ptr<PayloadPool> payload_pool,
+        fastrtps::rtps::RTPSParticipant* rtps_participant,
+        const bool repeater /* = false */)
+    : BaseWriter(participant_id, topic, payload_pool)
+    , rtps_participant_(rtps_participant)
+    , repeater_(repeater)
+{
+    // Do nothing
+}
+
+MultiWriter::~MultiWriter()
+{
+    // Lock so no other operations is taking place
+    std::unique_lock<WritersMapType> lock(writers_map_);
+
+    // Disable every inside writer and kill it
+    for(auto& writer : writers_map_)
+    {
+        writer.second->disable();
+        delete writer.second;
+    }
+
+    logInfo(DDSROUTER_RTPS_WRITER, "Deleting MultiWriter created in Participant " <<
+            participant_id_ << " for topic " << topic_);
+}
+
+void MultiWriter::enable_() noexcept
+{
+    std::shared_lock<WritersMapType> lock(writers_map_);
+    for(auto& writer : writers_map_)
+    {
+        writer.second->enable();
+    }
+}
+
+void MultiWriter::disable_() noexcept
+{
+    std::shared_lock<WritersMapType> lock(writers_map_);
+    for(auto& writer : writers_map_)
+    {
+        writer.second->disable();
+    }
+}
+
+bool MultiWriter::exist_partition_(const types::SpecificWriterQoS& data_qos)
+{
+    std::shared_lock<WritersMapType> lock(writers_map_);
+    return writers_map_.find(data_qos) != writers_map_.end();
+}
+
+QoSSpecificWriter* MultiWriter::get_writer_or_create_(const types::SpecificWriterQoS& data_qos)
+{
+    // NOTE: it uses unique lock because it may change the database, and there is no way
+    // to do so if taking share and unique must be done.
```

Comment:
I would say this protection is not necessary, as `write` is called guarded from `BaseWriter`.

Reply:
Good point. It is not needed indeed.
However I would prefer the mutex here than in BasicWriter (grano fino) so probably the BasicWriter mutex should disappear.

Reply:
You are probably right.